### ISSUE TITLE
`-fsanitize=undefined` does not detect uninitialized lvalue

### ISSUE THREAD

**Issue by vinc17fr** [llvm/llvm-project#201850]
ISO C23 6.3.2.1p2 says: "If the lvalue designates an object of automatic storage duration that could have been declared with the register storage class (never had its address taken), and that object is uninitialized (not declared with an initializer and no assignment to it has been performed prior to use), the behavior is undefined."

But on the following C code compiled with Clang 22.1.6 and `-fsanitize=undefined` under Debian/unstable, I do not get any warning.

```c
int main(void)
{
  int x;
  return x;
}
```

Note that this should not be confused with "MemorySanitizer: use-of-uninitialized-value" obtained by using `-fsanitize=memory`, because the following code also triggers a MemorySanitizer warning while this is not undefined behavior (the condition "never had its address taken" is not satisfied).

```c
int main(void)
{
  int x;
  (void) &x;
  return x;
}
```

Said otherwise, one cannot distinguish these two cases while there is a difference for ISO C23.

**Comment by zwuis** [comment]
What checks are enabled is clearly documented in <https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#available-checks>. Do you want `-fsanitize=undefined` to enable `-fsanitize=memory`?

**Comment by vinc17fr** [comment]
No, on the contrary. I want `-fsanitize=undefined` to detect undefined behavior due to the use of uninitialized data. Using `-fsanitize=memory` would be incorrect as it warns on correct code.

**Comment by efriedma-quic** [comment]
-fsanitize=undefined is not, and never will be, a tool to detect general uninitialized memory access; this is impossible to do in general in an ABI-compatible way, and anything involving pointers doesn't fit into the existing infrastructure.

We could add a check specifically to detect uses of uninitialized non-address-taken variables; that doesn't require any global state, just a local boolean to track whether the variable is initialized.  That said, I'm not sure it's worth implementing: we already have other tools for dealing with uninitialized variables.  It could be useful in situations where -fsanitize=memory isn't usable, I guess.

**Comment by AaronBallman** [comment]
> We could add a check specifically to detect uses of uninitialized non-address-taken variables;

Aren't there ways to initialize a non-address-taken variable outside of the base C language, like via linker scripts, inline assembly, etc? Those would be edge cases, but it would be odd for this code to trigger UBSan:

```c
int main() {
  int x;
  asm ("movl %1, %%eax;"
       "movl %%eax, %0;"
       :"=r"(x)
       :"r"(10)
       :"%eax"
  );
  return x;
}
```

or is the frontend smart enough to know that `x` has been initialized there?

**Comment by efriedma-quic** [comment]
We currently don't have any code in clang CodeGen which tries to track if a variable is "initialized" in the sense mentioned here.  Presumably we'd define it in a way that interacts reasonably with inline asm.

**Comment by devsw-prayas** [comment]
Opened a PR to implement this: llvm/llvm-project#207529
